src/store.c

```diff
@@ -290,47 +357,48 @@ static bool writer_open(
     return false;
 }
 
-static struct encoder writer_begin(
-        struct rill_store *store,
-        const struct vals *vals,
-        struct indexer *indexer)
+static void writer_flush_indices(
+    struct rill_store *store,
+    struct indexer *indexer_a,
+    struct indexer *indexer_b)
 {
-    size_t len = sizeof(*vals) + sizeof(vals->data[0]) * vals->len;
-    assert(store->head->vals_off + len < store->vma_len);
+    const size_t indexer_a_size = indexer_cap(indexer_a->len);
+    const size_t indexer_b_size = indexer_cap(indexer_b->len);
 
-    memcpy(store->vals, vals, len);
+    // TODO: this should replicate what used to be done for indices
+    // before they were moved to the top of the file, though has less
+    // meaning to do so now. Will have to recheck with Remi.
 
-    store->head->data_off = store->head->vals_off + len;
-    store->data = (void *) ((uintptr_t) store->vma + store->head->data_off);
+    indexer_write(indexer_a,
+                  store->index,
+                  store->vma_len - (indexer_a_size + indexer_b_size));
 
-    return store_encoder(store, indexer);
+    indexer_write(indexer_b,
+                  store->index_b,
+                  store->vma_len - (indexer_b_size));
 }
 
 static void writer_close(
-        struct rill_store *store, struct indexer *indexer, size_t len)
+    struct rill_store *store, size_t len)
 {
     if (len) {
         assert(len <= store->vma_len);
 
-        store->head->index_off = len;
-        store->index = (void *) ((uintptr_t) store->vma + store->head->index_off);
-        len += indexer_write(indexer, store->index, store->vma_len - len);
-
+        // Indicate that the file has been fully written and is ready for
+        // use. An additional sync is required for the stamp to ensure that the
+        // data is...
+        // - ... properly persisted before we delete it (durability)
+        // - ... only persisted after all the data has been persisted (ordering)
```

Comment:
this comment should remain where it was or moved to the header of the function. Note that there are two calls to `fdatasync` and this comment was set to be next to the write to `stamp` as explains what stamp does and is used for.

Reply:
Sorry about that. Fixed.